src/spatialdata/_core/query/relational_query.py

```diff
@@ -247,33 +311,25 @@ def _inner_join_spatialelement_table(
                     )
                     continue
 
-                mask = table_instance_key_column.isin(element_indices)
-                masked_table_instance_key_column = table_instance_key_column[mask]
-
-                masked_element = (
-                    element.iloc[mask_values, :]
-                    if len(mask_values := masked_table_instance_key_column.values) != 0
-                    else None
-                )
+                masked_element = _get_masked_element(element_indices, element, table_instance_key_column, match_rows)
                 element_dict[element_type][name] = masked_element
 
-                if joined_indices is None:
-                    joined_indices = masked_table_instance_key_column.index
-                else:
-                    # in place append does not work with pd.Index
-                    joined_indices = joined_indices.append(masked_table_instance_key_column.index)
+                joined_indices = _get_joined_table_indices(
+                    joined_indices, element_indices, table_instance_key_column, match_rows
+                )
             else:
                 warnings.warn(
                     f"The element `{name}` is not annotated by the table. Skipping", UserWarning, stacklevel=2
                 )
                 element_dict[element_type][name] = None
                 continue
+
     joined_table = table[joined_indices, :].copy() if joined_indices is not None else None
     return element_dict, joined_table
 
 
 def _left_exclusive_join_spatialelement_table(
-    element_dict: dict[str, dict[str, Any]], table: AnnData
+    element_dict: dict[str, dict[str, Any]], table: AnnData, match_rows: str
```

Comment:
why str and not `Literal['left' ,'no']`?

Reply:
resolved